java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -959,53 +964,68 @@ class DatabaseImpl(
         schema: Schema,
         db: SQLiteDatabase,
         counters: Counters? = null
-    ): TypeId = schemaMutex.withLock {
-        schemaTypeMap[schema.hash]?.let {
-            counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_HIT)
-            return@withLock it
-        }
-        counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_MISS)
-
-        return db.transaction {
-            counters?.increment(DatabaseCounters.INSERT_ENTITY_TYPE_ID)
+    ): TypeId = db.transaction {
+        var cacheHit = false
+        val result = schemaTypeMap.updateAndGet { currentMap ->
+            currentMap[schema.hash]?.let {
+                cacheHit = true
+                return@updateAndGet currentMap
+            }
+            cacheHit = false
```

Comment:
It looks like cacheHit is always false here anyway

Reply:
the way atomic.update methods work is they re-run the block until it agrees with itself (ie. if two threads are mutating it at once, they both repeat their mutations until the end result has settled).. so we need to make the internals idempotent (and thus - re-set the flag to false each time it's run, same reason for the CONFLICT_IGNORE)

Reply:
For reference: https://github.com/Kotlin/kotlinx.atomicfu/blob/master/atomicfu/src/commonMain/kotlin/kotlinx/atomicfu/AtomicFU.common.kt#L122-L128